crates/astria-telemetry/src/display.rs

```diff
@@ -26,13 +37,75 @@ pub fn hex<T: AsRef<[u8]> + ?Sized>(bytes: &T) -> Hex<'_> {
 /// A newtype wrapper of a byte slice that implements [`std::fmt::Display`].
 ///
 /// To be used in tracing contexts. See the [`self::hex`] utility.
-pub struct Hex<'a>(pub &'a [u8]);
+pub struct Hex<'a>(&'a [u8]);
 
-impl<'a> std::fmt::Display for Hex<'a> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+impl<'a> Display for Hex<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
         for byte in self.0 {
             f.write_fmt(format_args!("{byte:02x}"))?;
         }
         Ok(())
     }
 }
+
+/// Format `bytes` as lower-cased hex.
+///
+/// # Example
+/// ```
+/// use astria_telemetry::display;
+/// let signature = vec![1u8, 2, 3, 4, 5, 6, 7, 8];
+/// tracing::info!(signature = %display::hex(&signature), "received signature");
+/// ```
+pub fn json<T>(serializable: &T) -> Json<'_, T>
+where
+    T: serde::Serialize,
+{
+    Json(serializable)
+}
+
+/// A newtype wrapper of a serializable type that implements [`std::fmt::Display`].
+///
+/// To be used in tracing contexts. See the [`self::json`] utility.
+///
+/// # Panics
+/// The type must not contain non-utf8 fields, nor can any of the type's fields or variants
+/// have [`Serialize`] implementations that are fallible. The [`Display`] implementation will
+/// panic otherwise.
+pub struct Json<'a, T>(&'a T);
+
+impl<'a, T> Display for Json<'a, T>
+where
+    T: serde::Serialize,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        struct WriterFormatter<'a, 'b: 'a> {
+            inner: &'a mut Formatter<'b>,
+        }
+
+        impl<'a, 'b> io::Write for WriterFormatter<'a, 'b> {
+            fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+                // NOTE: This implementation is lifted straight from serde_json.
+                // Safety: the serializer below only emits valid utf8 when using
+                // the default formatter.
+                let s = unsafe { str::from_utf8_unchecked(buf) };
```

Comment:
is this guaranteed safe? what does `non-utf8` fields mean, since this is serializing something? if this function already returns a result why not just return an error instead of panicking?

Reply:
I have added links to serde_json's `impl Display for Value`, as that's lifted from there. But yeah - it's guaranteed safe since the serde_json serializer will return an error if it encounters non-utf8 strings.